### PR TITLE
CORE-4190: Fix memoryview refhandling

### PR DESCRIPTION
**Pull request by klaussfreire**
Turns out deallocation from `__dealloc__` was having issues. Python 3.4 and above support using a regular finalizer from cython now, `__del__`, so added that one. To keep it functional in older python versions, kept the dealloc around, but it should do nothing if `__del__` ran.

Also some other fixes:
* cast also reshapes, so check ndim
* buffer = memoryview in py3, so `type(x) is buffer` checks work as expected

### CODE REVIEW

**Review by gefarion — APPROVED** [review]

File: sharedbuffers/mapped_struct.py [hunk 1]
```diff
@@ -132,19 +132,28 @@
     long = int
     basestring = (bytes, str)
 
+_memoryview = cython.declare(object)
+_memoryview = memoryview
```
Comment (chpedersen) [inline sharedbuffers/mapped_struct.py:136]: This is to save us the `__Pyx_GetModuleGlobalName` stuff, right?

File: sharedbuffers/mapped_struct.py [hunk 2]
```diff
@@ -132,19 +132,28 @@
...
+            a = _memoryview(a)
+        if a.format != 'B' or a.ndim != 1:
+            a = a.cast('B')
+        return a
     else:
         return buffer(a)
 
 if cython.compiled:
     # Compatibility fix for cython >= 0.23, which no longer supports "buffer" as a built-in type
+    buffer = cython.declare(object, buffer)  # lint:ok
```
Comment (chpedersen) [inline sharedbuffers/mapped_struct.py:152]: I understand the `cython.declare(object` but why initialize it to "itself"?

File: sharedbuffers/mapped_struct.py [hunk 3]
```diff
@@ -1752,7 +1772,11 @@ def _init(self):
             if self.pybuf.buf != cython.NULL:
```
Comment (chpedersen) [inline sharedbuffers/mapped_struct.py:1769]: How can this be != null here?

File: sharedbuffers/mapped_struct.py [hunk 4]
```diff
@@ -3476,6 +3505,13 @@ def __dealloc__(self):
                 PyBuffer_Release(cython.address(self.pybuf))  # lint:ok
                 self.pybuf.buf = cython.NULL
 
+    def __del__(self):
+        # For python 3.4+, in earlier versions it will not be called and we'll depend on dealloc
+        if cython.compiled:
+            if self.pybuf.buf != cython.NULL:
+                PyBuffer_Release(cython.address(self.pybuf))  # lint:ok
+                self.pybuf.buf = cython.NULL
+
```
Comment (chpedersen) [inline sharedbuffers/mapped_struct.py:3511]: I commented out these two `__del__` and it works :shrug:

**Review by chpedersen — APPROVED** [review]

File: sharedbuffers/mapped_struct.py (hunk 1 above)
Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:136]: Aha

File: sharedbuffers/mapped_struct.py (hunk 2 above)
Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:152]: I just want a native global pointer to `buffer`, and I'm calling it `buffer`

File: sharedbuffers/mapped_struct.py (hunk 3 above)
Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:1769]: If it's being reinitialized (reset)

File: sharedbuffers/mapped_struct.py (hunk 4 above)
Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:3511]: Sure. you'll have a memory leak though.

File: sharedbuffers/mapped_struct.py (hunk 2 above)
Comment (chpedersen) [reply sharedbuffers/mapped_struct.py:152]: I mean, you could just do `buffer = cython.declare(object)`, you are setting it right below this

Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:152]: Both syntaxes do the same, but that one would leave it uninitialized until it's initialized explicitly. The one I used initializes it from the start, right there in the declaration, which allows those variables to be const (not that this one is).

**Pull request merged**